components/electric/lib/electric/replication/postgres/logical_replication_producer.ex

```diff
@@ -339,37 +374,28 @@ defmodule Electric.Replication.Postgres.LogicalReplicationProducer do
   def handle_demand(incoming_demand, %{demand: pending_demand} = state) do
     state = %{state | demand: incoming_demand + pending_demand}
 
-    dispatch_events(state, [])
+    dispatch_events(state)
   end
 
-  # When we're done exhausting demand, emit events.
-  defp dispatch_events(%{demand: 0} = state, events) do
-    emit_events(state, events)
+  defp dispatch_events(%{demand: demand, queue_len: queue_len} = state)
+       when demand == 0 or queue_len == 0 do
+    {:noreply, [], state}
   end
 
-  defp dispatch_events(%{demand: demand, queue: queue} = state, events) do
-    case :queue.out(queue) do
-      # If the queue has events, recurse to accumulate them
-      # as long as there is demand.
-      {{:value, event}, queue} ->
-        state = %{state | demand: demand - 1, queue: queue}
-
-        dispatch_events(state, [event | events])
+  defp dispatch_events(%{demand: demand, queue: queue, queue_len: queue_len} = state)
+       when demand >= queue_len do
+    queue |> :queue.last() |> ack(state)
 
-      # When the queue is empty, emit any accumulated events.
-      {:empty, queue} ->
-        state = %{state | queue: queue}
-
-        emit_events(state, events)
-    end
+    state = %{state | queue: :queue.new(), queue_len: 0, demand: demand - queue_len}
+    {:noreply, :queue.to_list(queue), state}
   end
 
-  defp emit_events(state, []) do
-    {:noreply, [], state}
-  end
+  defp dispatch_events(%{demand: demand, queue: queue, queue_len: queue_len} = state) do
+    {to_emit, queue_remaining} = :queue.split(demand, queue)
+    to_emit |> :queue.last() |> ack(state)
 
-  defp emit_events(state, events) do
-    {:noreply, Enum.reverse(events), state}
+    state = %{state | queue: queue_remaining, queue_Len: queue_len - demand, demand: 0}
```

Comment:
Typo that should lead to a runtime error, possibly means that this code path is not covered by test (although this is caught by Dialyzer)
```suggestion
    state = %{state | queue: queue_remaining, queue_len: queue_len - demand, demand: 0}
```

Reply:
Yeah, I had to stop working on this abruptly yesterday and pushed what I had ready by then. This condition requires a specific test setup to cover. I think having dialyzer complain about it is good enough.